README.md
|      |      |
| ---- | ---- |
|  作者    |  [gaoyong](mailto:gaoyong06@qq.com) |
|  创建时间    | 2020-11-04 13:40:00      |
# gleam

一个简单、轻量、可靠的Flutter UI 组件库

## 介绍

在业务开发中经常会有一些常用的组件，将这些常用的组件标准化后，有了这个库

希望gleam是一个简单、轻量、可靠的Flutter UI 组件库

## 安装

1. 在项目的pubspec.yaml中在增加下面配置：gleam: ^1.0.0
2. 第二步：然后在项目根目录(pubspec.yaml 所在的目录)执行命令：flutter pub get

## 介绍

##### BottomSheet 底部菜单弹窗

​	底部弹出的菜单

###### 示意图
<img src="https://github.com/gaoyong06/gleam/blob/master/doc/1604912567034167.gif" style="zoom: 40%;"/>


|  <img src="https://ww1.sinaimg.cn/mw690/00019562gy1gkj1gbkf8kj20u01t0tbu.jpg" style="zoom:40%"/>    | <img src="https://ww1.sinaimg.cn/mw690/00019562gy1gkj1gq2kx0j20u01t077m.jpg" style="zoom:40%"/>     |   <img src="https://ww1.sinaimg.cn/mw690/00019562gy1gkj1h1fc9jj20u01t0424.jpg" style="zoom:40%"/>   |   <img src="https://ww1.sinaimg.cn/mw690/00019562gy1gkj1h9l63pj20u01t0tce.jpg" style="zoom:40%"/>   |   <img src="https://ww1.sinaimg.cn/mw690/00019562gy1gkj1hrawwyj20u01t0q61.jpg" style="zoom:40%"/>   |
| ---- | ---- | ---- | ---- | ---- |
|  基础用法    |   展示取消按钮   |   展示描述信息   |  选项状态    |  自定义面板   |


调用showGleamBottomSheet函数后弹出的底部菜单弹窗，其中操作项使用**BottomSheetAction**组件

###### 	参数说明

| 参数名称      | 说明                                                         | 类型               | 默认值 |
| ------------- | ------------------------------------------------------------ | ------------------ | ------ |
| context       | 上下文                                                       | BuildContext       | -      |
| actions       | 操作项列表                                                   | List<Widget>       | -      |
| contentWidget | 内容区域Widget (如果设置contentWidget，则actions、description、cancelText、onCancelTap的设置不生效) | Widget             | -      |
| description   | 顶部显示的描述(可以是文字或者Widget)                         | dynamic            | -      |
| cancelText    | 取消文案（没有设置则不显示“取消”按钮）                       | String             | -      |
| onCancelTap   | 取消按钮点击回调方法                                         | GestureTapCallback | -      |

**BottomSheetAction**组件属性说明

| 属性名称      | 说明                         | 类型               | 默认值 |
| ------------- | ---------------------------- | ------------------ | ------ |
| item          | 操作项(可以是文字或者Widget) | dynamic            | -      |
| height        | 操作项高度                   | double             | 50     |
| onTap         | 操作项点击回调方法           | GestureTapCallback | -      |
| hiddenDivider | 是否隐藏操作项下方分割线     | bool               | false  |

###### 	代码演示

```dart
//基础用法
showGleamBottomSheet(context, actions: [
  BottomSheetAction('选项一', onTap: () {
    showToast('选项一');
  }),
  BottomSheetAction('选项二', hiddenDivider: true, onTap: () {
    showToast('选项二');
  }),
]);

//展示取消按钮
showGleamBottomSheet(
  context,
  actions: [
    BottomSheetAction('选项一', onTap: () {
      showToast('选项一');
    }),
    BottomSheetAction(
      '选项二',
      hiddenDivider: true,
      onTap: () {
        showToast('选项二');
      },
    ),
  ],
  cancelText: '取消',
  onCancelTap: () {
    showToast('取消');
  },
);

//展示描述信息
showGleamBottomSheet(
  context,
  actions: [
    BottomSheetAction('选项一', onTap: () {
      showToast('选项一');
    }),
    BottomSheetAction(
      '选项二',
      hiddenDivider: true,
      onTap: () {
        showToast('选项二');
      },
    ),
  ],
  description: Container(
    padding: EdgeInsets.only(top: 20.0, bottom: 20.0),
    height: 60.0,
    child: Text(
      '这是描述信息',
      style: Style.normal15Color999999,
    ),
  ),
  cancelText: '取消',
  onCancelTap: () {
    showToast('取消');
  },
);

//选项状态
showGleamBottomSheet(
  context,
  actions: [
    BottomSheetAction(
      Text(
        '着色选项',
        style: Style.normal15ColorFF0000,
      ), onTap: () {
        showToast('着色状态');
      }),
    BottomSheetAction(
      Text(
        '禁用选项',
        style: Style.normal15Color999999,
      ),
      onTap: null,
    ),
    BottomSheetAction(
      CupertinoActivityIndicator(),
      hiddenDivider: true,
      onTap: () {
        showToast('loading...');
      },
    ),
  ],
  cancelText: '取消',
  onCancelTap: () {
    showToast('取消');
  },
);

//自定义面板
showGleamBottomSheet(
  context,
  contentWidget: Container(
    height: 200.0,
    child: Stack(
      children: [
        Positioned(
          top: 0,
          right: 0,
          child: IconButton(
            icon: Icon(Icons.close),
            color: AppColors.clC8C9CC,
            onPressed: () {
              Navigator.of(context).pop();
            },
          ),
        ),
        Column(
          children: [
            SizedBox(
              height: 16.0,
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                Text(
                  "标题",
                  style: Style.bold15Color333333,
                )
              ],
            ),
            Expanded(
              child: Center(
                child: Text('内容'),
              ),
            ),
          ],
        ),
      ],
    ),
  ),
              );
```




##### Dialog 弹出框

​	弹出模态框，常用于消息提示、消息确认、在当前页面内完成特定的交互操作

##### Empty 空状态

​	空状态时的占位提示

##### FilledButton 填充按钮

​	填充按钮的文字为白色，背景为按钮的颜色

##### GleamOutlineButton 细边框按钮

​	细边框按钮的文字,和边框为按钮的颜色，背景为白色

##### GleamTextButton 文字按钮

​	文字按钮只有一个文字，该文字可以点击

##### GleamTextField 文本输入框

​	提供文本输入框右侧尾随的Widget扩展

##### GleamIcon 图标

​	提供Icon及Icon上红点及消息未读数显示

##### GleamImage 图片

​	提供多种图片来源网络图片、本地图片、资源图片，支持圆角、支持占位图、失败图

##### ImagePicker 照片选择

​	从手机相册选择照片

##### Placeholder 占位状态

​	占位状态时的组件	

##### Search 搜索

​	搜索输入框

##### ShareSheet 分享面板

​	底部弹起的分享面板，用于展示各分享渠道对应的操作按钮，不含具体的分享逻辑。

##### Tag 标签

​	文字，或者图标+

## 依赖

## 打包
  export https_proxy=http://127.0.0.1:7890 http_proxy=http://127.0.0.1:7890 all_proxy=socks5://127.0.0.1:7890

  set http_proxy=http://127.0.0.1:7890
  set https_proxy=http://127.0.0.1:7890
  
  flutter packages pub publish --dry-run
  flutter packages pub publish --server=https://pub.dartlang.org

## 团队成员
  - [@gaoyong](https://github.com/gaoyong06)
  - [@liuxiaobina](https://github.com/liuxiaobina)
  - [@lee980](https://github.com/lee980)

  













CHANGELOG.md
<!--
/*
 * @Author: gaoyong06@qq.com
 * @Date: 2020-11-06 14:08:23 
 * @Last Modified by: gaoyong06@qq.com
 * @Last Modified time: 2020-11-06 14:08:23
 */
-->
## [1.0.0] - 小目标

* 基础组件：8
* 表单组件：15
* 反馈组件：9
* 展示组件：15
* 导航组件：8
* 业务组件：11

* 共计：66
* 业务组件以外共计：55

## [1.0.x] - 计划做的事情.

* TODO: GleamButton 无需设置width
* TODO: GleamTextField 增加输入验证规则
* TODO: GleamTextField 增加显示字数统计
* TODO: GleamTextField 增加clearable 是否启用清除图标，点击清除图标后会清空输入框
* TODO: GleamButton 中的文字按钮，参考一下系统 TextButton 的实现
* TODO: Tag 增加单选，多选
* TODO: 增加 Mock请求 
* TODO: 了解一下CupertinoAlertDialog




example/lib/main.dart
/*
 * @Author: gaoyong06@qq.com
 * @Date: 2020-11-06 14:08:23 
 * @Last Modified by: gaoyong06@qq.com
 * @Last Modified time: 2020-11-06 14:08:23
 */
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_screenutil/flutter_screenutil.dart';
import 'package:gleam/gleam.dart';
import 'package:gleam/style/app_colors.dart';
import 'package:gleam/style/style.dart';
import 'package:gleam_example/pages/bottom_sheet_page.dart';
import 'package:gleam_example/pages/cell_page.dart';
import 'package:gleam_example/pages/empty_page.dart';
import 'package:gleam_example/pages/gleam_button_page.dart';
import 'package:gleam_example/pages/gleam_check_box_page.dart';
import 'package:gleam_example/pages/gleam_dialog_page.dart';
import 'package:gleam_example/pages/gleam_icon_page.dart';
import 'package:gleam_example/pages/gleam_image_page.dart';
import 'package:gleam_example/pages/gleam_text_field_page.dart';
import 'package:gleam_example/pages/layout_page.dart';
import 'package:gleam_example/pages/search_page.dart';
import 'package:gleam_example/pages/share_sheet_page.dart';
import 'package:gleam_example/pages/tag_page.dart';
import 'package:oktoast/oktoast.dart';

void main() {
  WidgetsFlutterBinding.ensureInitialized();

  //强制竖屏
  SystemChrome.setPreferredOrientations(
      [DeviceOrientation.portraitUp, DeviceOrientation.portraitDown]);
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return OKToast(
      child: MaterialApp(
        debugShowCheckedModeBanner: false, //去掉Flutter右上角的debug
        title: 'gleam example',
        theme: ThemeData(
          primarySwatch: Colors.blue,
          visualDensity: VisualDensity.adaptivePlatformDensity,
        ),
        home: MyHomePage(title: 'gleam example'),
      ),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  //标题样式
  TextStyle textStyle = TextStyle(
    color: Color(0XFF323233),
    fontSize: 14.0,
    fontWeight: FontWeight.w600,
  );

  //标题
  Widget _title(String text) {
    return Container(
      padding: EdgeInsets.fromLTRB(16.0, 24.0, 0, 8),
      child: Text(text ?? '', style: Style.normal14Color66455A64),
    );
  }

  //按钮
  Widget _gleamButton(String text, [Function() onPressed]) {
    Widget _content = Row(
      mainAxisSize: MainAxisSize.max,
      mainAxisAlignment: MainAxisAlignment.spaceBetween,
      children: [
        Expanded(
          child: Text(
            text,
            textAlign: TextAlign.start,
            style: TextStyle(
              color: Color(0XFF323233),
              fontSize: 14.0,
              fontWeight: FontWeight.w600,
            ),
          ),
        ),
        Icon(
          Icons.chevron_right,
          size: 18.0,
          color: Color(0XFFB6C3D2),
        )
      ],
    );

    return GleamButton(
      style: GleamButtonStyle.flatButton,
      type: ButtonType.defaulted,
      margin: EdgeInsets.only(top: 10.0),
      content: _content,
      borderRadius: 99.0,
      fillColor: AppColors.clF7F8FA,
      onPressed: onPressed,
    );
  }

  @override
  Widget build(BuildContext context) {
    //设置适配尺寸 (填入设计稿中设备的屏幕尺寸) 此处假如设计稿是按iPhone6的尺寸设计的(iPhone6 750*1334)
    ScreenUtil.init(context,
        designSize: Size(750 / 2, 1334 / 2), allowFontScaling: false);
    return Scaffold(
      backgroundColor: Color(0XFFFFFFFF),
      appBar: AppBar(
        title: Text(
          widget.title,
          style: Style.bold18ColorFFFFFF,
        ),
      ),
      body: ListView(
        padding: EdgeInsets.symmetric(vertical: 0, horizontal: 16.0),
        children: [
          SizedBox(
            height: 40.0,
          ),
          Row(
            children: [
              GleamImage(
                  image:
                      'https://tvax2.sinaimg.cn/crop.0.0.1079.1079.1024/00019562ly1gkffsfazmwj20u00tz0uf.jpg?KID=imgbed,tva&Expires=1604654948&ssig=66QSdllaXt',
                  width: 60,
                  height: 60,
                  shape: BoxShape.circle),
              SizedBox(
                width: 16.0,
              ),
              Text(
                'gleam',
                style: TextStyle(fontSize: 30.0, fontWeight: FontWeight.w600),
              )
            ],
          ),
          SizedBox(
            height: 16.0,
          ),
          Text(
            '一个简单、轻量、可靠的Flutter UI 组件库',
            style: Style.normal14Color66455A64,
          ),
          SizedBox(
            height: 16.0,
          ),
          _title('基础组件'),
          _gleamButton(
            'GleamButton 按钮',
            () {
              Navigator.push(
                context,
                MaterialPageRoute(builder: (context) => GleamButtonPage()),
              );
            },
          ),
          _gleamButton(
            'Cell 单元格',
            () {
              Navigator.push(
                context,
                MaterialPageRoute(builder: (context) => CellPage()),
              );
            },
          ),
          _gleamButton(
            'GleamIcon 图标',
            () {
              Navigator.push(
                context,
                MaterialPageRoute(builder: (context) => GleamIconPage()),
              );
            },
          ),
          _gleamButton(
            'GleamImage 图片',
            () {
              Navigator.push(
                context,
                MaterialPageRoute(builder: (context) => GleamImagePage()),
              );
            },
          ),
          _gleamButton(
            'Layout 布局#',
            () {
              Navigator.push(
                context,
                MaterialPageRoute(builder: (context) => LayoutPage()),
              );
            },
          ),
          _gleamButton(
            'Popup弹出层#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Style 内置样式#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Toast 轻提示#',
            () {
              showToast("未完成");
            },
          ),
          _title('表单组件'),
          _gleamButton(
            'Calendar 日历#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'GleamCheckbox 复选框',
            () {
              Navigator.push(
                context,
                MaterialPageRoute(builder: (context) => GleamCheckBoxPage()),
              );
            },
          ),
          _gleamButton(
            'DatetimePicker 时间选择#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'GleamTextField 输入框#',
            () {
              Navigator.push(
                context,
                MaterialPageRoute(builder: (context) => GleamTextFieldPage()),
              );
            },
          ),
          _gleamButton(
            'Form 表单#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Numberkeyboard 数字键盘#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'PasswordInput 密码输入框#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Picker 选择器#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Radio 单选框#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Rate 评分#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Search 搜索',
            () {
              Navigator.push(
                context,
                MaterialPageRoute(builder: (context) => SearchPage()),
              );
            },
          ),
          _gleamButton(
            'Slider 滑块#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Stepper 步进器#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Switch 开关#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Uploader 文件上传#',
            () {
              showToast("未完成");
            },
          ),
          _title('反馈组件'),
          _gleamButton(
            'BottomSheet 底部菜单弹窗',
            () {
              Navigator.push(
                context,
                MaterialPageRoute(builder: (context) => BottomSheetPage()),
              );
            },
          ),
          _gleamButton(
            'GleamDialog 弹出框',
            () {
              Navigator.push(
                context,
                MaterialPageRoute(builder: (context) => GleamDialogPage()),
              );
            },
          ),
          _gleamButton(
            'DropdownMenu 下拉菜单#',
            () {
              Navigator.push(
                context,
                MaterialPageRoute(builder: (context) => GleamDialogPage()),
              );
            },
          ),
          _gleamButton(
            'Loading 加载#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Notify 消息通知#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Overlay 遮罩层#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'PullRefresh 下拉刷新#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'ShareSheet 分享面板#',
            () {
              Navigator.push(
                context,
                MaterialPageRoute(builder: (context) => ShareSheetPage()),
              );
            },
          ),
          _gleamButton(
            'SwipeCell 滑动单元格#',
            () {
              showToast("未完成");
            },
          ),
          _title('展示组件'),
          _gleamButton(
            'Badge 微标#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Circle 环形进度条#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Collapse 折叠面板#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'CountDown 倒计时#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Divider 分割线#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Empty 空状态',
            () {
              Navigator.push(
                context,
                MaterialPageRoute(builder: (context) => EmptyPage()),
              );
            },
          ),
          _gleamButton(
            'ImagePreview 图片预览#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Lazyload 懒加载#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'List 列表#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'NoticeBar 通知栏#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Popover 气泡弹出窗#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Progress 进度条#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Skeleton 骨架屏#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Steps 步骤条#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Sticky 粘性布局#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Swipe 轮播#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Tag 标签',
            () {
              Navigator.push(
                context,
                new MaterialPageRoute(builder: (context) => TagPage()),
              );
            },
          ),
          _title('导航组件'),
          _gleamButton(
            'Grid 宫格#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'IndexBar 索引栏#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'NavBar 导航栏#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Pagination 分页#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Sidebar 侧边导航#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Tab 标签页#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'Tabbar 标签栏#',
            () {
              showToast("未完成");
            },
          ),
          _gleamButton(
            'TreeSelect 分类选择#',
            () {
              showToast("未完成");
            },
          ),
          _title('其他'),
          _gleamButton(
            '联系作者#',
            () {
              showToast("未完成");
            },
          ),
          SizedBox(
            height: 40.0,
          ),
          Column(
            children: [
              Text(
                'Made With ❤️ by gaoyong',
                style: Style.normal12Color455A64,
              ),
              SizedBox(
                height: 4.0,
              ),
              Text(
                'Version 1.0.0',
                style: Style.normal12Color455A64,
              ),
              SizedBox(
                height: 4.0,
              ),
            ],
          ),
          SizedBox(
            height: 40.0,
          ),
        ],
      ),
    );
  }
}
